File: src/main/java/com/github/ushiosan23/networkutils/download/DownloadElement.java
package com.github.ushiosan23.networkutils.download;

import com.github.ushiosan23.networkutils.http.HttpRequestAction;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URI;
import java.net.URL;
import java.net.http.HttpHeaders;
import java.util.Calendar;
import java.util.OptionalLong;

/**
 * Download element object
 */
public final class DownloadElement extends BaseDownload {

	/* ---------------------------------------------------------
	 *
	 * Properties
	 *
	 * --------------------------------------------------------- */

	/**
	 * Target url to download
	 */
	private final URI downloadURL;

	/**
	 * Save download headers (download once)
	 */
	private HttpHeaders downloadHeaders;

	/**
	 * Download status event
	 */
	private DownloadStatusEvent lastStatus;

	/**
	 * Download thread
	 */
	private Thread downloadThread;

	/* ---------------------------------------------------------
	 *
	 * Constructors
	 *
	 * --------------------------------------------------------- */

	/**
	 * Create download element from uri object
	 *
	 * @param url Target url to download
	 */
	public DownloadElement(@NotNull URI url) {
		// Check url scheme
		if (!url.getScheme().equals("http") && !url.getScheme().equals("https"))
			throw new IllegalArgumentException(String.format("\"%s\" has not valid http scheme.", url));

		// Set properties
		downloadURL = url;
		isPaused = false;
		isCancelled = false;
		isFinished = false;
		downloadBytes = 0;
		isIndefinite = false;
	}

	/**
	 * Create download element from url string
	 *
	 * @param url Target url to download
	 */
	public DownloadElement(String url) {
		this(URI.create(url));
	}

	/* ---------------------------------------------------------
	 *
	 * Methods
	 *
	 * --------------------------------------------------------- */

	/**
	 * Check if download exists
	 *
	 * @return Request result
	 */
	public boolean downloadExists() {
		return HttpRequestAction.exists(downloadURL);
	}

	/**
	 * Get download headers
	 *
	 * @return {@link HttpHeaders} request headers result
	 */
	public HttpHeaders downloadHeaders() {
		if (downloadHeaders == null)
			downloadHeaders = HttpRequestAction.getHeaders(downloadURL);

		return downloadHeaders;
	}

	/**
	 * Check download size
	 *
	 * @return Download size in bytes. If download size is -1 this download is indefinite
	 */
	public long downloadSize() {
		HttpHeaders localHeaders = downloadHeaders();
		OptionalLong optionalLong = localHeaders.firstValueAsLong("content-length");

		if (optionalLong.isPresent()) {
			isIndefinite = false;
			return optionalLong.getAsLong();
		} else {
			isIndefinite = true;
			return -1L;
		}
	}

	/**
	 * Start download method
	 */
	public void startDownload() {
		downloadThread = new Thread(downloadRunnable());
		downloadThread.start();
	}

	/**
	 * Waits for this thread to die.
	 *
	 * @throws InterruptedException if any thread has interrupted the current thread.
	 *                              The interrupted status of the current thread is cleared when this exception is thrown.
	 */
	public void join() throws InterruptedException {
		if (downloadThread != null)
			downloadThread.join();
	}

	/**
	 * Get download file name. Return url if download is not a file
	 *
	 * @return Download file name
	 */
	@NotNull
	public String getDownloadFileName() {
		String path = downloadURL.getPath();
		int lastIndex = path.lastIndexOf("/");
		String result;

		if (lastIndex != -1) {
			result = path.substring(lastIndex + 1);
		} else {
			result = path;
		}

		return result.isEmpty() ? path : result;
	}

	/* ---------------------------------------------------------
	 *
	 * Internal methods
	 *
	 * --------------------------------------------------------- */

	/**
	 * Runnable process download
	 *
	 * @return {@link Runnable} Task process download
	 */
	@NotNull
	@Contract(pure = true)
	private Runnable downloadRunnable() {
		// Runnable download
		return () -> {
			try {
				downloadProcess();
			} catch (Exception e) {
				e.printStackTrace();
			}
		};
	}

	/**
	 * Download process
	 *
	 * @throws Exception Any download error
	 */
	private void downloadProcess() throws Exception {
		int downloadRead;
		// Initialize properties
		File tmpFile = generateTemporalFile(downloadURL);
		FileOutputStream fOutput = new FileOutputStream(tmpFile);
		BufferedInputStream buffered = new BufferedInputStream(downloadURL.toURL().openStream());
		lastStatus = DownloadStatusEvent.getDefault(DownloadElement.this);
		downloadBuffer = new byte[1024];
		// Configure event
		lastStatus.isIndefinite = isIndefinite();
		lastStatus.totalSize = downloadSize();

		// Iterate all download data
		while (
			(downloadRead = buffered.read(downloadBuffer, 0, downloadBuffer.length)) != -1 &&
				!isCancelled() &&
				!isPaused()
		) {
			// Write download data
			fOutput.write(downloadBuffer);
			// Change download status
			lastStatus.downloaded += downloadRead;
			if (!lastStatus.isIndefinite)
				lastStatus.downloadProgress = lastStatus.downloaded * 100f / lastStatus.totalSize;
			// Fire events
			fireListener(lastStatus.cloneSelf());
		}

		// Close resources
		fOutput.flush();
		fOutput.close();
		buffered.close();
		// Configure status
		lastStatus.isCancelled = isCancelled();
		lastStatus.isPaused = isPaused();
		lastStatus.isFinished = !isPaused() && !isCancelled();
		lastStatus.tmpFile = lastStatus.isFinished ? tmpFile : null;

		fireListener(lastStatus.cloneSelf());
	}

	@NotNull
	private static String getFileDownload(@NotNull URI uri) {
		String path = uri.getPath();
		int lastIndex = path.lastIndexOf("/");
		String result;

		if (lastIndex != -1) {
			result = path.substring(lastIndex);
		} else {
			result = path;
		}

		return result.isEmpty() ? path : result;
	}

	@NotNull
	private static File generateTemporalFile(URI uri) throws IOException {
		long timeInMillis = Calendar.getInstance().getTimeInMillis();
		String hexString = Long.toHexString(timeInMillis);
		String file = getFileDownload(uri);

		return File.createTempFile(String.format("%s.%s", file, hexString), ".tmpdownload");
	}

}

File: src/main/kotlin/com/github/ushiosan23/networkutils/download/DownloadElement.kt
package com.github.ushiosan23.networkutils.download

import com.github.ushiosan23.networkutils.extensions.toHexString
import com.github.ushiosan23.networkutils.extensions.toURI
import com.github.ushiosan23.networkutils.http.HttpRequestActions
import java.io.BufferedInputStream
import java.io.File
import java.net.URI
import java.net.URL
import java.net.URLConnection
import java.net.http.HttpHeaders
import java.util.*

/**
 * Download class element
 */
class DownloadElement : IDownloadElement {

	/* ---------------------------------------------------------
	 *
	 * Public properties
	 *
	 * --------------------------------------------------------- */

	/**
	 * Determine if download was cancelled
	 */
	@Volatile
	var isCancelled: Boolean = false
		private set

	/**
	 * Determine if download was paused
	 */
	@Volatile
	var isPaused: Boolean = false
		private set

	/* ---------------------------------------------------------
	 *
	 * Internal properties
	 *
	 * --------------------------------------------------------- */

	/**
	 * Download url element
	 */
	private val downloadURI: URI

	/**
	 * Download url element
	 */
	private val downloadURL: URL

	/**
	 * URL connection
	 */
	private var connection: URLConnection? = null

	/**
	 * Temporal download file
	 */
	private var tmpOutputFile: File? = null

	/**
	 * Download status object
	 */
	@Volatile
	private var lastDownloadStatus: DownloadStatus? = null

	/* ---------------------------------------------------------
	 *
	 * Constructors
	 *
	 * --------------------------------------------------------- */

	/**
	 * URI constructor
	 *
	 * @param url Target url to download. Valid shcemes (http, https)
	 *
	 * @throws IllegalArgumentException if url scheme is not valid
	 */
	constructor(url: URI) {
		// Validate shceme
		if (url.scheme != "https" && url.scheme != "http")
			throw IllegalArgumentException("\"$url\" has not valid http scheme.")
		// Set properties
		downloadURI = url
		downloadURL = downloadURI.toURL()
	}

	/**
	 * String constructor
	 *
	 * @param url Target url string
	 *
	 * @throws IllegalArgumentException if url scheme is not valid
	 * @throws java.net.URISyntaxException if url is not valid
	 */
	constructor(url: String) : this(url.toURI())

	/**
	 * URL constructor
	 *
	 * @param url Target url object
	 *
	 * @throws IllegalArgumentException if url scheme is not valid
	 * @throws java.net.URISyntaxException if url is not valid
	 */
	constructor(url: URL) : this(url.toURI())

	/* ---------------------------------------------------------
	 *
	 * Implemented methods
	 *
	 * --------------------------------------------------------- */

	/**
	 * Check if url exists
	 *
	 * @return [Boolean] download status
	 */
	override fun exists(): Boolean =
		HttpRequestActions.urlExists(downloadURI)

	/**
	 * Get download url headers
	 *
	 * @return [HttpHeaders] result headers
	 *
	 * @see HttpRequestActions.urlHeaders
	 */
	override fun getHeaders(): HttpHeaders? = if (exists())
		HttpRequestActions.urlHeaders(downloadURI)
	else
		null

	/**
	 * Get document total size
	 *
	 * @return [Long] with total document size or `-1` if size is not defined
	 *
	 * @see HttpRequestActions.urlContentLength
	 */
	override fun getContentLength(): Long =
		HttpRequestActions.urlContentLength(downloadURI)

	/**
	 * Download file chunk by chunk
	 *
	 * @param info Download status
	 */
	override fun download(info: (status: DownloadStatus) -> Unit) {
		// Generate download status and temporal file
		tmpOutputFile = generateTemporalFile(downloadURI)
		lastDownloadStatus = DownloadStatus()

		// Start download section
		try {
			// Create local vars
			var iterationRead = 0
			val outputStream = tmpOutputFile!!.outputStream()
			val httpStream = BufferedInputStream(downloadURL.openStream())
			val buffer = ByteArray(1024)
			// Set initial download status values
			lastDownloadStatus!!.totalBytes = getContentLength()
			lastDownloadStatus!!.currentBytes = 0L

			// Read data
			do {
				// Check if download is cancelled
				if (isCancelled) {
					outputStream.flush()
					outputStream.close()
					httpStream.close()
					throw InterruptedException("Download was cancelled")
				}

				// Store read buffer size
				iterationRead = httpStream.read(buffer, 0, buffer.size)

				// Check download iteration
				if (iterationRead != -1) {
					outputStream.write(buffer, 0, iterationRead)
					lastDownloadStatus!!.currentBytes += iterationRead
					info.invoke(lastDownloadStatus!!)
				} else {
					outputStream.flush()
					outputStream.close()
					httpStream.close()
					// Set new status data
					lastDownloadStatus!!.outputFile = tmpOutputFile!!
					info.invoke(lastDownloadStatus!!)
				}

			} while (iterationRead != -1)


		} catch (err: Exception) {
			err.printStackTrace()
			// Delete temporal file
			if (tmpOutputFile?.exists()!!)
				tmpOutputFile?.delete()
			// Set download failed
			lastDownloadStatus!!.hasError = true
			lastDownloadStatus!!.downloadError = err
			// Invoke method
			info.invoke(lastDownloadStatus!!)
		}

	}

	/**
	 * Cancel download
	 */
	override fun cancel() = synchronized(this) {
		isCancelled = true
	}

	/**
	 * Pause current download
	 *
	 * @param info download status
	 */
	override fun pause(info: (status: DownloadStatus) -> Unit) = synchronized(this) {
		super.pause(info)
		/*lastDownloadStatus?.let(info::invoke)
		isPaused = true*/
	}

	/**
	 * Resume download
	 */
	override fun resume() = synchronized(this) {
		super.resume()
		/*isPaused = false*/
	}

	/**
	 * Companion object element
	 */
	companion object {

		/**
		 * Get current time in milliseconds
		 *
		 * @return [Long] with date milliseconds
		 */
		private fun getCurrentTime(): Long = Calendar.getInstance().timeInMillis

		/**
		 * Get file url name
		 *
		 * @param url Target url to check
		 *
		 * @return [String] with filename result
		 */
		private fun getURIFileDownload(url: URI): String {
			val path = url.path
			val lastIndex = path.lastIndexOf("/")
			val result = if (lastIndex != -1) path.substring(lastIndex) else path

			return if (result.isEmpty()) path else result
		}

		/**
		 * Generate temporal file
		 *
		 * @param url Target url generator file
		 */
		private fun generateTemporalFile(url: URI): File {
			val hexString = getCurrentTime().toHexString()
			val file = getURIFileDownload(url)

			return File.createTempFile("${file}.${hexString}", ".tmpdownload")
		}

	}

}
